internal/pkg/agent/application/enroll/enroll.go

```diff
@@ -0,0 +1,567 @@
+		}
+
+		cfg.Server.TLS.CAs = cas
+
+		var cAuth tlscommon.TLSClientAuth
+		cfg.Server.TLS.ClientAuth = &cAuth
+		if err := cfg.Server.TLS.ClientAuth.Unpack(clientAuth); err != nil {
+			return nil, errors.New(err, "failed to unpack --fleet-server-client-auth", errors.TypeConfig)
+		}
+	}
+
+	if localFleetServer {
+		cfg.Client.Transport.Proxy.Disable = true
+		cfg.Server.InternalPort = internalPort
+	}
+
+	if err := cfg.Valid(); err != nil {
+		return nil, errors.New(err, "invalid enrollment options", errors.TypeConfig)
+	}
+	return cfg, nil
+}
+
+func CreateAgentConfig(agentID string, pc map[string]interface{}, headers map[string]string, staging string) map[string]interface{} {
+	agentConfig := map[string]interface{}{
+		"id": agentID,
+	}
+
+	if len(headers) > 0 {
+		agentConfig["headers"] = headers
+	}
+
+	if staging != "" {
+		staging := fmt.Sprintf("https://staging.elastic.co/%s-%s/downloads/", release.Version(), staging[:8])
+		agentConfig["download"] = map[string]interface{}{
+			"sourceURI": staging,
+		}
+	}
+
+	for k, v := range pc {
+		agentConfig[k] = v
+	}
+
+	return agentConfig
+}
+
+func SafelyStoreAgentInfo(s saver, reader io.Reader) error {
+	var err error
+	signal := make(chan struct{})
+	backExp := backoff.NewExpBackoff(signal, 100*time.Millisecond, 3*time.Second)
+
+	for i := 0; i <= maxRetriesstoreAgentInfo; i++ {
+		backExp.Wait()
+		err = storeAgentInfo(s, reader)
+		if !errors.Is(err, filelock.ErrAppAlreadyRunning) {
+			break
+		}
+	}
+
+	close(signal)
+	return err
+}
+
+func LoadPersistentConfig(pathConfigFile string) (map[string]interface{}, error) {
+	persistentMap := make(map[string]interface{})
+	rawConfig, err := config.LoadFile(pathConfigFile)
+	if os.IsNotExist(err) {
+		return persistentMap, nil
+	}
+	if err != nil {
+		return nil, errors.New(err,
+			fmt.Sprintf("could not read configuration file %s", pathConfigFile),
+			errors.TypeFilesystem,
+			errors.M(errors.MetaKeyPath, pathConfigFile))
+	}
+
+	pc := &struct {
+		Headers        map[string]string                      `json:"agent.headers,omitempty" yaml:"agent.headers,omitempty" config:"agent.headers,omitempty"`
+		LogLevel       string                                 `json:"agent.logging.level,omitempty" yaml:"agent.logging.level,omitempty" config:"agent.logging.level,omitempty"`
+		MonitoringHTTP *monitoringConfig.MonitoringHTTPConfig `json:"agent.monitoring.http,omitempty" yaml:"agent.monitoring.http,omitempty" config:"agent.monitoring.http,omitempty"`
+	}{
+		MonitoringHTTP: monitoringConfig.DefaultConfig().HTTP,
+	}
+
+	if err := rawConfig.UnpackTo(&pc); err != nil {
+		return nil, err
+	}
+
+	if pc.LogLevel != "" {
+		persistentMap["logging.level"] = pc.LogLevel
+	}
+
+	if pc.MonitoringHTTP != nil {
+		persistentMap["monitoring.http"] = pc.MonitoringHTTP
+	}
+
+	return persistentMap, nil
+}
+
+func delay(ctx context.Context, d time.Duration) {
+	t := time.NewTimer(rand.N(d))
+	defer t.Stop()
+	select {
+	case <-ctx.Done():
+	case <-t.C:
+	}
```

Comment:
I know this code is not new, but just a result of refactoring, but there is an opportunity for a nitpicky improvement, if you wish to take it 😄.

```suggestion
	select {
	case <-ctx.Done():
	case <-time.After(rand.N(d)):
	}
```